optimade/server/models/structures.py

```diff
@@ -1,564 +1,718 @@
+    - :val:`unknown_positions`: This flag MUST be present if at least one component of the :property:`cartesian_site_positions` list of lists has value :val:`null`.
+    - :val:`assemblies`: This flag MUST be present if the property `assemblies`_ is present.
+
+-  **Examples**: A structure having unknown positions and using assemblies: :val:`["assemblies", "unknown_positions"]`""",
     )
 
+    @validator("elements", whole=False)
+    def element_must_be_chemical_symbol(cls, v):
+        assert (
+            v in CHEMICAL_SYMBOLS
+        ), f"Only chemical symbols are allowed, you passed: {v}"
+        return v
+
+    @validator("elements", whole=True)
+    def elements_must_be_alphabetical(cls, v):
+        assert sorted(v) == v, f"elements must be sorted alphabetically, but is: {v}"
+        return v
+
+    @validator("elements_ratios", whole=True)
+    def ratios_must_sum_to_one(cls, v):
+        assert (
+            sum(v) == 1
+        ), f"elements_ratios MUST sum to 1 within floating point accuracy. It sums to: {sum(v)}"
+        return v
+
+    @validator("chemical_formula_reduced", "chemical_formula_hill")
+    def no_spaces_in_reduced(cls, v):
+        assert " " not in v, f"Spaces are not allowed, you passed: {v}"
+        return v
+
+    @validator("dimension_types", whole=True)
+    def must_be_of_length_three(cls, v):
+        assert len(v) == 3, f"MUST be of length 3, but is of length: {len(v)}"
+        for dimension in v:
+            assert dimension in {0, 1}, f"MUST be either 0 or 1, you passed: {v}"
+        return v
+
+    @validator("lattice_vectors", always=True, whole=True)
+    def required_if_dimension_types_has_one(cls, v, values):
+        if 1 in values.get("dimension_types") and v is None:
+            raise AssertionError(
+                f"lattice_vectors is REQUIRED, since dimension_types is not [0, 0, 0] but is {values.get('dimension_types')}"
+            )
+        assert_msg = f"MUST be a an 3 x 3 array (list of 3 lists of 3 floats), found instead: {v}"
+        assert len(v) == 3, assert_msg
+        return v
+
+    @validator("lattice_vectors", "cartesian_site_positions")
+    def sites_must_have_length_three(cls, v):
+        assert len(v) == 3, f"MUST be a list of length 3. {v} has length {len(v)}."
+        return v
+
+    @validator("nsites")
+    def validate_nsites(cls, v, values):
+        assert v == len(
+            values.get("cartesian_site_positions", [])
+        ), f"nsites (value: {v}) MUST equal length of cartesian_site_positions (value: {len(values.get('cartesian_site_positions', []))})"
+        return v
+
+    @validator("species_at_sites", whole=True)
+    def validate_species_at_sites(cls, v, values):
+        assert len(v) == values.get(
+            "nsites", 0
+        ), f"Number of species_at_sites (value: {len(v)}) MUST equal number of sites (value: {values.get('nsites', 0)})"
+        assert len(set(v)) == len(v), "Species in species_at_sites MUST be unique"
```

Comment:
Unless I'm mistaken, this is incorrect.

Reply:
One other thing; in the current structure mapper we currently project the results of the mongo query onto all the OPTiMaDe keys. I have my documents stored with data saved under the `attributes` key as they were made directly from the models here. Do we want to support this? Would mean the `fields` set has to include both `"attributes"` and all of the requested provider fields prepended with `attributes`...

Reply:
> One other thing; in the current structure mapper we currently project the results of the mongo query onto all the OPTiMaDe keys. I have my documents stored with data saved under the `attributes` key as they were made directly from the models here. Do we want to support this? Would mean the `fields` set has to include both `"attributes"` and all of the requested provider fields prepended with `attributes`...

This I consider an implementation issue, i.e., not something that we should solve here with the "test" server.
People may store their OPTiMaDe databases in different ways, they should write make their mapper accordingly.
So for this, I do not think it is an issue - if I understood your question here correct? :)

Reply:
> Unless I'm mistaken, the second assert is incorrect (and unnecessary). The spec states "Each species MUST have a unique name." but not each entry in this list must be unique. Admittedly the spec is a bit confusing...

Yeah, I may have missed something here. Was getting a bit confused at some point with all the validations. Will fix

Reply:
I guess my problem is that the structure mapper doesn't actually receive the structure from the collection, but instead only receives the optimade fields + top level provider fields, so we are forcing providers to store optimade keys at the top level. I think it would be nice if we could be a bit more flexible, perhaps with an optional config entry for provider_optimade_key or some such so it can look for the optimade fields inside e.g. `doc['optimade']` or in my case `doc['attributes']` to avoid polluting the toplevel of any existing database's schema. Obviously these are all implementation dependent but I think its nice to be flexible where its trivial to be so! Thoughts?

Reply:
> I guess my problem is that the structure mapper doesn't actually receive the structure from the collection, but instead only receives the optimade fields + top level provider fields, so we are forcing providers to store optimade keys at the top level. I think it would be nice if we could be a bit more flexible, perhaps with an optional config entry for provider_optimade_key or some such so it can look for the optimade fields inside e.g. `doc['optimade']` or in my case `doc['attributes']` to avoid polluting the toplevel of any existing database's schema. Obviously these are all implementation dependent but I think its nice to be flexible where its trivial to be so! Thoughts?

Fine for me, but personally, I would just write a completely new mapper. Ideally _only_ reuse the models and write everything else anew, to fit my personal requirements and my particular database setup.
However, if you already have a working mapper for this particular case, I don't see why we can't buff up the test server with some more advanced possibilities.

Reply:
Writing a new mapper is fine (and its what I've done), but its a very small change in the `MongoCollection` that I'd like to accommodate to avoid duplicating all of the test server code for collections. I'll knock something together as a separate PR perhaps!

Reply:
> Writing a new mapper is fine (and its what I've done), but its a very small change in the `MongoCollection` that I'd like to accommodate to avoid duplicating all of the test server code for collections. I'll knock something together as a separate PR perhaps!

Sounds great! :) I think we should start keeping the changes to the server out of this PR now - I'm even starting to regret the "fix" of "response_fields" - but in the end justify it with: "The changes we make in the models that affect the test server, should not break the test server - if it does, we should fix it".
The problem being that the response_fields was not working even prior to this PR's changes, hehe...

Reply:
agreed :P